### ISSUE TITLE
add --max-active-adapters argument

### ISSUE THREAD

**Issue by flozi00** [predibase/lorax#92]
### Feature request

as discussed here: https://discord.com/channels/1174495433565945916/1176984269558652998/1179487788266172508

### Motivation

If there are to many requests for different adapters you are running out of memory, since the actual system only checks the memory for batch sizes and not concurrent adapters.

### Your contribution

I looked at the effort it would take to add this feature and realized it was beyond my ability with Rust

**Comment by tgaddair** [comment]
To clarify, we do have the `--max-active-adapters` argument, but it doesn't account for the fact that adapters can be of varying sizes. Ideally, I would prefer something like `--max-active-adapter-memory` that sets aside a reservation of GPU memory for adapters, such that we would place adapters on GPU up to the memory limit, rather than up to the fixed max adapters. Is that also what you had in mind for this?

**Comment by flozi00** [comment]
Okay, then it was an misunderstanding from my side.
How is the system working? 
If the active adapters limit is reached, are the requests added to queue or return an specific error code?

**Comment by tgaddair** [comment]
Ah, okay, I think you're right @flozi00. Looks like `--max-active-adapters` is only wired up when launching the router individually, but isn't being plumbed through to the `lorax-launcher` that starts both the server and the router. We should definitely fix this (same for `--adapter-cycle-time-s`.

**Comment by flozi00** [comment]
```
error: unexpected argument '--max-active-adapters' found

  tip: a similar argument exists: '--max-batch-prefill-tokens'

Usage: lorax-launcher <--model-id <MODEL_ID>|--adapter-id <ADAPTER_ID>|--source <SOURCE>|--adapter-source <ADAPTER_SOURCE>|--revision <REVISION>|--validation-workers <VALIDATION_WORKERS>|--sharded <SHARDED>|--num-shard <NUM_SHARD>|--quantize <QUANTIZE>|--dtype <DTYPE>|--trust-remote-code|--max-concurrent-requests <MAX_CONCURRENT_REQUESTS>|--max-best-of <MAX_BEST_OF>|--max-stop-sequences <MAX_STOP_SEQUENCES>|--max-input-length <MAX_INPUT_LENGTH>|--max-total-tokens <MAX_TOTAL_TOKENS>|--waiting-served-ratio <WAITING_SERVED_RATIO>|--max-batch-prefill-tokens <MAX_BATCH_PREFILL_TOKENS>|--max-batch-total-tokens <MAX_BATCH_TOTAL_TOKENS>|--max-waiting-tokens <MAX_WAITING_TOKENS>|--hostname <HOSTNAME>|--port <PORT>|--shard-uds-path <SHARD_UDS_PATH>|--master-addr <MASTER_ADDR>|--master-port <MASTER_PORT>|--huggingface-hub-cache <HUGGINGFACE_HUB_CACHE>|--weights-cache-override <WEIGHTS_CACHE_OVERRIDE>|--disable-custom-kernels|--cuda-memory-fraction <CUDA_MEMORY_FRACTION>|--json-output|--otlp-endpoint <OTLP_ENDPOINT>|--cors-allow-origin <CORS_ALLOW_ORIGIN>|--watermark-gamma <WATERMARK_GAMMA>|--watermark-delta <WATERMARK_DELTA>|--ngrok|--ngrok-authtoken <NGROK_AUTHTOKEN>|--ngrok-edge <NGROK_EDGE>|--env|--download-only>

```

`
docker run --pull always -v ./data:/data  --gpus all -d --shm-size 1g -p 8080:80 ghcr.io/predibase/lorax:latest --model-id mistralai/Mistral-7B-Instruct-v0.1 --cuda-memory-fraction 0.5 --max-total-tokens 8192 --max-batch-prefill-tokens 8000 --max-input-length 7000 --dtype float16 --max-active-adapters 1

`

**Comment by tgaddair** [comment]
Hmm, this looks to be an issue with the Docker build process, possibly a bad cache hit. I can verify that when I install in my dev environment, the new arg is available.

cc @noyoshi @gyanesh-mishra any ideas?

**Comment by tgaddair** [comment]
I suspect the issue here is coming from `cargo chef`: https://www.lpalmieri.com/posts/fast-rust-docker-builds/

**Comment by tgaddair** [comment]
@flozi00, I just landed #97 in part as an attempt to force a cache miss in Docker on most of these components. Looks like it worked after testing against the latest Docker image. Can you try pulling and testing again?

**Comment by flozi00** [comment]
container is starting now as it should
